workflow/controller/hooks.go

```diff
@@ -0,0 +1,109 @@
+package controller
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/argoproj/argo-workflows/v3/errors"
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	"github.com/argoproj/argo-workflows/v3/util/template"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	"github.com/argoproj/argo-workflows/v3/workflow/templateresolution"
+)
+
+func IsExitHook(hookName wfv1.LifecycleEvent) bool {
+	return hookName == wfv1.ExitLifecycleEvent
+}
+
+func (woc *wfOperationCtx) executeWfLifeCycleHook(ctx context.Context, tmplCtx *templateresolution.Context) error {
+	if woc.wf.Spec.Hooks != nil && len(woc.wf.Spec.Hooks) > 0 {
+		for hookName, hook := range woc.wf.Spec.Hooks {
+			//exit hook will be execute in runOnExitNode
+			if IsExitHook(hookName) {
+				continue
+			}
+			// Replace hook's parameters
+			hookBytes, err := json.Marshal(hook)
+			if err != nil {
+				return errors.InternalWrapError(err)
```

Comment:
is there a strong reason to wrap like this? would it be good to use `fmt.Errorf` and add some diagnostics?